test/e2e/kubelet_deletes.go

```diff
@@ -0,0 +1,146 @@
+
+const (
+	// Interval to poll /runningpods on a node
+	pollInterval = 1 * time.Second
+)
+
+// waitTillKubeletDeletesPods waits till pods with UID in targetPodUIDs set
+// are visible on the kubelet's /runningpods endpoint.
+func waitTillKubeletDeletesPods(restClient *client.Client, targetPodUIDs util.StringSet, nodeName string, timeout time.Duration) error {
+	return wait.Poll(pollInterval, timeout, func() (bool, error) {
+		pods, err := GetKubeletPods(restClient, nodeName)
+		if err != nil {
+			return false, err
+		}
+
+		alive := 0
+		for _, pod := range pods.Items {
+			if targetPodUIDs.Has(string(pod.UID)) {
+				alive += 1
+			}
+		}
+		if alive > 0 {
+			Logf("Waiting on kubelet to delete %d pods on node %q", alive, nodeName)
+			return false, nil
+		}
+		return true, nil
+	})
+}
+
+var _ = Describe("Clean up pods on node", func() {
+	var numNodes int
+	framework := NewFramework("kubelet-delete")
+
+	BeforeEach(func() {
+		nodes, err := framework.Client.Nodes().List(labels.Everything(), fields.Everything())
+		expectNoError(err)
+		numNodes = len(nodes.Items)
+	})
+
+	type DeleteTest struct {
+		podsPerNode int
+		timeout     time.Duration
+	}
+
+	deleteTests := []DeleteTest{
+		{podsPerNode: 10, timeout: 3 * time.Minute},
+	}
+
+	for _, itArg := range deleteTests {
+		name := fmt.Sprintf(
+			"kubelet should be able to delete %d pods per node in %v.", itArg.podsPerNode, itArg.timeout)
+		It(name, func() {
+			var labelSelector labels.Selector
+			nodeNames := util.NewStringSet()
+
+			totalPods := itArg.podsPerNode * numNodes
+
+			By(fmt.Sprintf("Creating a RC of %d pods and wait until all pods of this RC are running", totalPods))
+			rcName := "cleanup" + strconv.Itoa(totalPods) + "-" + string(util.NewUUID())
+
+			Expect(RunRC(RCConfig{
+				Client:    framework.Client,
+				Name:      rcName,
+				Namespace: framework.Namespace.Name,
+				Image:     "gcr.io/google_containers/pause:go",
+				Replicas:  totalPods,
+			})).NotTo(HaveOccurred())
+
+			labelSelector = labels.Set(map[string]string{"name": rcName}).AsSelector()
+			pods, err := framework.Client.Pods(framework.Namespace.Name).List(labelSelector, fields.Everything())
+			expectNoError(err)
+
+			// Record the nodes on which the pods of this RC are scheduled, as
+			// well as UIDs of the pods.
+			desired := util.NewStringSet()
+			for _, pod := range pods.Items {
+				nodeNames.Insert(pod.Spec.NodeName)
+				desired.Insert(string(pod.UID))
+			}
+
+			// Perform a sanity check so that we know all desired pods are
+			// running on the nodes according to kubelet.
+			seen := util.NewStringSet()
+			for _, name := range nodeNames.List() {
+				Logf("Checking pods on node %v via /runningpods endpoint", name)
+				runningPods, err := GetKubeletPods(framework.Client, name)
+				expectNoError(err)
+				for _, pod := range runningPods.Items {
+					seen.Insert(string(pod.UID))
+				}
+			}
+			Expect(seen.IsSuperset(desired)).To(BeTrue(), "All pods should be running on nodes")
```

Comment:
See above.  I don't think that this is strictly speaking true.  Pods could have failed and been moved. At the very least please have the text description reflect that this is perhaps unusual but not theoretically incorrect, in case someone needs to debug this in future.